pennylane/gradients/pulse_generator_gradient.py

```diff
@@ -0,0 +1,229 @@
+        The ``k``\ th entry of the returned ``tuple`` has the shape ``(2**N, 2**N, *par_shape)``
+        where ``N`` is the number of qubits the evolution acts on and ``par_shape`` is the
+        shape of the ``k``\ th parameter of the evolution.
+
+    The effective generator can be computed from the derivative of the unitary
+    matrix corresponding to the full time evolution of a pulse:
+
+    .. math::
+
+        \Omega_k = U(\theta)^\dagger \frac{\partial}{\partial \theta_k} U(\theta)
+
+    Here :math:`U(\theta)` is the unitary matrix of the time evolution due to the pulse
+    and :math:`\theta` are the variational parameters of the pulse.
+
+    See the documentation of pulse_generator for more details and a mathematical derivation.
+    """
+
+    def _compute_matrix(*args):
+        """Parametrized computation of the matrix for the given pulse ``op``."""
+        return op(*args, t=op.t, **op.odeint_kwargs).matrix()
+
+    def _compute_matrix_split(*args):
+        """Parametrized computation of the matrix for the given pulse ``op``.
+        Return the real and imaginary parts separately."""
+        mat = _compute_matrix(*args)
+        return mat.real, mat.imag
+
+    # Compute the Jacobian of _compute_matrix, giving the Jacobian of the real and imag parts
+    # The output is a tuple, with one entry per parameter, each of which has the axes
+    # (mat_dim, mat_dim, *parameter_shape)
+    jac_real, jac_imag = jax.jacobian(_compute_matrix_split, argnums=0)(op.data)
+
+    # Compute the matrix of the pulse itself and conjugate it. Skip the transposition of the adjoint
+    # The output has the shape (mat_dim, mat_dim)
+    U_dagger = qml.math.conj(_compute_matrix([qml.math.detach(d) for d in op.data]))
+
+    # Compute U^\dagger @ \partial U / \partial \theta_k
+    # For each entry ``j`` in the tuple ``jac``,
+    # 1. Contract U_dagger with j along the axes [0], [0]. This effectively transposes
+    #    U_dagger, which we skipped above.
+    # 2. After contraction, the axes are (mat_dim, mat_dim, *parameter_shape).
+    # 3. Move the first two axis to the last two positions.
+    moveax = partial(qml.math.moveaxis, source=(0, 1), destination=(-2, -1))
+    return tuple(
+        moveax(qml.math.tensordot(U_dagger, j_r + 1j * j_i, axes=[[0], [0]]))
+        for j_r, j_i in zip(jac_real, jac_imag)
+    )
+
+
+def _one_parameter_paulirot_coeffs(generators, num_wires):
+    r"""Compute the Pauli coefficients of the one-parameter group generators that reproduce
+    the partial derivatives of a parametrized evolution. The coefficients correspond
+    to the decomposition into (rescaled) Pauli word generators as used by ``PauliRot``
+    gates.
+
+    Args:
+        generators (tuple[tensor_like]): Generators of the one-parameter groups that
+            reproduce the partial derivatives of the parametrized evolution of interest.
+        num_wires (int): Number of wires the parametrized evolution acts on.
```

Comment:
I'm a bit torn here. It's a private method, so it's fine not to do validation/pretty error messages, I'd say.
The `num_wires` information is readily available when we call this function, and while it is a very cheap operation, I'd rather provide the additional information from where we have it anyways, instead of computing it anew.

Reply:
I definitely see your point. I don't feel strongly about this :-) Resolving this.